Debug

sreynen opened the pull request:
This fixes a few different bugs, more detail in the individual commits.

Commit:
whitespace cleanup

Commit:
add headers for easier debugging

Commit:
fix transactions, immutable field map

Commit:
allow non-transactions tables to run

Commit:
search disputes by received_date, not created_at

Commit:
case_number isn't actually a number

sreynen commented on lib/field-maps.js:
```diff
@@ -41,26 +41,164 @@ exports.transactions = ({
   getGateway,
-  ...fields
-}) => ({
-  ...fields,
```
Braintree API added more fields, which added more fields to the field map, which broke the import because those fields don't exist in the table schema. The fix is to explicitly state all the fields we care about, and ignore all other fields.

sreynen commented on lib/braintree.js:
```diff
@@ -34,9 +34,19 @@ module.exports = class {
+
+      if (responseObject == 'disputes') {
+        searchStream = this.client.dispute.search(search => (
+          isMin ? search.receivedDate().min(max) : search.receivedDate().max(min)
```
Braintree API doesn't allow searching disputes by createdAt() for some reason, so this searches by receivedDate() instead, which seems like the closest searchable field.

schuyler1d replied:
Seems like a good thing to add as a comment in the code.

sreynen commented on lib/table-schemas.js:
```diff
@@ -160,7 +160,7 @@ exports.disputes = SQL`
   amount_disputed DECIMAL(10,2) ENCODE RAW,
   amount_won DECIMAL(10,2) ENCODE RAW,
-  case_number DECIMAL ENCODE LZO,
+  case_number VARCHAR ENCODE LZO,
```
Case number values look like "CB190644814501".

schuyler1d commented on lib/braintree.js:
```diff
@@ -29,15 +28,25 @@ module.exports = class {
     const tmpPath = process.env.NODE_ENV !== 'development' ? '/tmp' : path.join(__dirname, '../tmp');
     const csvPath = path.join(tmpPath, csvName);
-    const writer = csvWriter({ separator: '|', sendHeaders: false });
+    const writer = csvWriter({ separator: '|' });
```
nit, but if this was a gotcha, then I'd rather it be an explicit choice even if it's the default, so maybe add/keep: `sendHeaders: true`

schuyler1d reviewed (approved):
:+1: 

Looking at the incremental changes this all looks fine.  I was not able to follow the `isMin` logic or figure out what to expect in the redshiftTimestamps.  Further confusing in that was that the API calls `.min(max)` and the inverse and that `writePromise` is called with both values here:
https://github.com/MoveOnOrg/braintree-to-redshift/blob/470456d989155837f360d59c599912cda796a216/lib/braintree.js#L95
In order to understand the code, maybe it would be helpful to change some variable names (e.g. responseObject => table, since that's what it seems to be) -- and maybe rather than pack 'more' into variables like redshiftTimestamps, we just add more vars or something.  

My comments were requests just for more inline comments -- your PR comments would also be great additions in the code -- personally, I find the 'why's of something strange to be the most useful things to put in code comments -- those strange edge cases are almost always hard-earned wisdom and useful context.

Commit:
added some inline comments

sreynen commented:
I added some comments on the parts I added and understand. I agree the other parts are unclear, but they're also unclear to me, so I don't think I should be composing the comments to explain them, as that might just be reinforcing misunderstandings.

I think that's a symptom of a problem we can't solve in this PR: as far as I can tell, the initial deploy for this never went through a code review.
